src/document/generator.ml

```diff
@@ -1783,26 +1799,152 @@ module Make (Syntax : SYNTAX) = struct
+          (Inline [ inline (Inline.Styled (`Bold, [ inline (Inline.Text t) ])) ])
+      in
+      let page_hierarchy { Odoc_model.Lang.Sidebar.page_name; pages } =
+        if pages = [] then []
+        else
+          let pages = List.map prepare pages in
+          let hierarchy =
+            Hierarchy.make pages |> Hierarchy.remove_common_root
+          in
+          let convert (content, id, is_highlighted) =
+            let url = Url.from_identifier ~stop_before:false id in
+            match url with
+            | Ok href ->
+                let target = InternalLink.Resolved href in
+                let content = Comment.link_content content in
+                let link = { InternalLink.target; content; tooltip = None } in
+                let attr = if is_highlighted then [ "current_unit" ] else [] in
+                block (Inline [ inline ~attr @@ Inline.InternalLink link ])
+            | Error _ ->
+                let content = Comment.link_content content in
+                (* let attr = if is_highlighted then [ "current_unit" ] else [] in *)
+                block (Inline content)
+          in
+
+          let pages = Hierarchy.to_sidebar convert hierarchy in
+          let pages = [ block (Block.List (Block.Unordered, [ pages ])) ] in
+          [ title @@ page_name ^ "'s Pages" ] @ pages
+      in
+      let rec concat_map acc f = function
+        | hd :: tl -> concat_map (List.rev_append (f hd) acc) f tl
+        | [] -> List.rev acc
+      in
+      let page_hierarchies = concat_map [] page_hierarchy v.pages in
+      let units =
+        let item id =
+          let id = (id :> Paths.Identifier.t) in
+          let href, name, is_highlighted =
+            ( Url.from_identifier ~stop_before:false id,
+              Paths.Identifier.name id,
+              Odoc_model.Paths.Identifier.equal id root_id )
+          in
+          match href with
+          | Error _ -> None
+          | Ok href ->
+              let target = InternalLink.Resolved href in
+              let content = [ inline @@ Text name ] in
+              let link = { InternalLink.target; content; tooltip = None } in
+              let attr = if is_highlighted then [ "current_unit" ] else [] in
+              let elem =
+                [ block (Inline [ inline ~attr @@ Inline.InternalLink link ]) ]
+              in
+              Some elem
+        in
+        List.map
+          (fun { Odoc_model.Lang.Sidebar.units; name } ->
+            let units = filter_map [] item units in
+            [ title name; block (List (Block.Unordered, units)) ])
+          v.libraries
+      in
+      let units = block (Block.List (Block.Unordered, units)) in
+      let units = [ title "Libraries"; units ] in
+      page_hierarchies @ units
+
+    let compilation_unit ?sidebar:sb (t : Odoc_model.Lang.Compilation_unit.t) =
       let url = Url.Path.from_identifier t.id in
+      let sidebar =
+        match sb with None -> None | Some sb -> Some ((sidebar t.id) sb)
```

Comment:
I think these parentheses are not doing anything. (Same below)

Reply:
I'm too used to ocamlformat being great at eliminating useless parentheses!

Reply:
Not this one, that would be illegal in 5.2 :police_car: